edb/edgeql/compiler/viewgen.py

```diff
@@ -997,50 +970,104 @@ def _compile_rewrites_for_stype(
     ir_set: irast.Set,
     anchors: RewriteAnchors,
     s_ctx: ShapeContext,
+    *,
+    already_defined_rewrites: Optional[
+        Mapping[sn.UnqualName, EarlyShapePtr]] = None,
     ctx: context.ContextLevel,
 ) -> Dict[sn.UnqualName, EarlyShapePtr]:
+    schema = ctx.env.schema
+
     path_id = ir_set.path_id
-    (subject_set, specified_set, old_set) = anchors
 
     res = {}
 
-    scls_pointers = stype.get_pointers(ctx.env.schema)
-    for pn, ptrcls in scls_pointers.items(ctx.env.schema):
-        if ptrcls.is_pure_computable(ctx.env.schema):
+    if stype in ctx.active_rewrites:
+        vn = stype.get_verbosename(ctx.env.schema)
+        raise errors.QueryError(
+            f"rewrite rule on {vn} is part of a rewrite rule cycle",
+        )
+
+    scls_pointers = stype.get_pointers(schema)
+    for pn, ptrcls in scls_pointers.items(schema):
+        if ptrcls.is_pure_computable(schema):
             continue
 
-        rewrite = ptrcls.get_rewrite(ctx.env.schema, kind)
+        rewrite = ptrcls.get_rewrite(schema, kind)
         if not rewrite:
             continue
+        rewrite_pointer = downcast(
+            s_pointers.Pointer, rewrite.get_subject(schema))
+        rewrite_source = downcast(
+            s_objtypes.ObjectType, rewrite_pointer.get_source(schema))
+
+        # get_rewrite searches in ancestors for rewrites, but if the rewrite
+        # for that ancestor has already been compiled, skip it to avoid
+        # duplicating work
+        if (
+            already_defined_rewrites
+            and (existing := already_defined_rewrites.get(pn))
+            and (existing[0].get_nearest_non_derived_parent(schema)
+                 == rewrite_pointer)
+        ):
+            continue
+
+        subject_set, specified_set, old_set = anchors
+
+        rewrite_view = view_scls
+        if rewrite_source != view_scls.get_nearest_non_derived_parent(schema):
+            # FIXME: Caching?
+            rewrite_view = downcast(
+                s_objtypes.ObjectType,
+                schemactx.derive_view(
+                    rewrite_source,
+                    exprtype=s_ctx.exprtype,
+                    ctx=ctx,
+                )
+            )
+            subject_set = setgen.class_set(
+                rewrite_view, path_id=subject_set.path_id, ctx=ctx)
+            if old_set:
+                old_set = setgen.class_set(
+                    rewrite_view, path_id=old_set.path_id, ctx=ctx)
+
+            ir_set = setgen.class_set(
+                rewrite_view, path_id=ir_set.path_id, ctx=ctx)
```

Comment:
Probably I should have put a comment here actually:
This gives the input sets the type of the relevant child, not of the parent it might be inheriting from.
Also `_normalize_view_ptr_expr` really wants its input to have a view type (that is, a subtype of the real object type), and not a material object type, so actually we make a new derived type